### PR TITLE
add route type enum

### LINKED ISSUE #254: [Feature Request] Refactor Code to include better types
**Current Behavior**

Right now, there are a lot of keys as `Strings` in hashmaps. And a lot of strings are being passed where enums would be a better choice.
I lot of problems are being faced here. https://github.com/sansyrox/robyn/pull/253

**Desired Behavior**

Refactor the codebase.

---
**Screenshots / Mockups**

**Alternatives**

---

### PR DESCRIPTION
**Pull request by suhailmalik07**
**Description**

This PR fixes #254 

I'm working on this. 

@sansyrox It will be great, If you can point me to the places from where you think code can be converted to enum and types?

### CODE REVIEW

File: src/routers/middleware_router.rs [hunk 1]
```diff
@@ -9,6 +9,8 @@ use matchit::Node;
 
 use anyhow::{bail, Error, Result};
 
+use super::router::RouteType;
```
Comment (sansyrox) [inline src/routers/middleware_router.rs:12]: I don't think `super` is required anymore in Rust.

File: src/routers/router.rs [hunk 2]
```diff
@@ -10,6 +10,22 @@ use matchit::Node;
...
+impl RouteType {
+    pub fn from_str(input: &str) -> RouteType {
+        match input {
+            "BEFORE_REQUEST" => RouteType::BeforeRequest,
+            "AFTER_REQUEST" => RouteType::AfterRequest,
+            _ => panic!("Invalid route type enum."),
+        }
+    }
+}
+
```
Comment (sansyrox) [inline src/routers/router.rs:28]: These routes are just for the middleware router and not the global router. So, I think something like `MiddlewareRouteType` would be more relevant.

Also, we usually create a `types` folder in the namespace. e.g. `routes/types.py`

**Review by sansyrox — COMMENTED** [review]
Hey @suhailmalik07 . 👋 

Thank you for the PR. 😄 

I have a few inline comments.

File: src/routers/middleware_router.rs (hunk 1 above)
Comment (suhailmalik07) [reply src/routers/middleware_router.rs]: another option is to write like this.
```
use crate::routers::types::MiddlewareRouteType;
```

File: src/routers/router.rs (hunk 2 above)
Comment (suhailmalik07) [reply src/routers/router.rs]: updated PR with given details. Thanks for pointing it out.

File: src/routers/middleware_router.rs (hunk 1 above)
Comment (sansyrox) [reply src/routers/middleware_router.rs]: @suhailmalik07 , yep. That is the preferred way.

File: src/routers/middleware_router.rs [hunk 3]
```diff
@@ -25,19 +27,21 @@ impl MiddlewareRouter {
     }
 
     #[inline]
-    fn get_relevant_map(&self, route: &str) -> Option<&RwLock<Node<(PyFunction, u8)>>> {
+    fn get_relevant_map(
+        &self,
+        route: MiddlewareRouteType,
```
Comment (sansyrox) [inline src/routers/middleware_router.rs:32]: just one last nit: `MiddlewareRoute` would make more sense than `MiddlewareRouteType`.

Reason:

If you ask the type of route then it should be `MiddlewareRoute` and not type = `MiddlewareRouteType.`

**Review by sansyrox — COMMENTED** [review]
Thanks again @suhailmalik07 for the quick updates. I just have two final suggestions before we can merge this PR 😄

File: src/routers/middleware_router.rs (hunk 1 above)
Comment (suhailmalik07) [reply src/routers/middleware_router.rs]: update with this.

File: src/routers/middleware_router.rs (hunk 3 above)
Comment (suhailmalik07) [reply src/routers/middleware_router.rs]: Sure, updated this also.

**Review by sansyrox — APPROVED** [review]
LGTM! Great work!

**Comment by sansyrox** [comment]
@suhailmalik07 , there are some conflicting files. Can you please rebase the latest changes?

**Comment by suhailmalik07** [comment]
> @suhailmalik07 , there are some conflicting files. Can you please rebase the latest changes?

Rebased 🙂

**Review by sansyrox — APPROVED** [review]
LGTM! 🚀 🟢

**Pull request merged**